crates/turbopack-ecmascript/src/analyzer/mod.rs

```diff
@@ -1007,58 +1214,375 @@ impl JsValue {
+    /// doesn't make sense. This is for optimization purposes.
+    pub fn is_unknown(&self) -> bool {
+        match self {
+            JsValue::Unknown(..) => true,
+            JsValue::Alternatives(_, list) => list.iter().any(|x| x.is_unknown()),
+            _ => false,
+        }
+    }
+
+    /// Checks if we known that the value is a string. Returns None if we
+    /// don't know. Returns Some if we know if or if not the value is a string.
+    pub fn is_string(&self) -> Option<bool> {
+        match self {
+            JsValue::Constant(ConstantValue::StrWord(..))
+            | JsValue::Constant(ConstantValue::StrAtom(..))
+            | JsValue::Concat(..) => Some(true),
+
+            JsValue::Constant(..)
+            | JsValue::Array(..)
+            | JsValue::Object(..)
+            | JsValue::Url(..)
             | JsValue::Module(..)
-            | JsValue::FreeVar(_) => true,
+            | JsValue::Function(..)
+            | JsValue::WellKnownObject(_)
+            | JsValue::WellKnownFunction(_) => Some(false),
+
+            JsValue::Not(..) => Some(false),
+            JsValue::Add(_, list) => any_if_known(list, JsValue::is_string),
+            JsValue::Logical(_, op, list) => match op {
+                LogicalOperator::And => {
+                    shortcircut_if_known(list, JsValue::is_truthy, JsValue::is_string)
+                }
+                LogicalOperator::Or => {
+                    shortcircut_if_known(list, JsValue::is_falsy, JsValue::is_string)
+                }
+                LogicalOperator::NullishCoalescing => {
+                    shortcircut_if_known(list, JsValue::is_not_nullish, JsValue::is_string)
+                }
+            },
+
+            JsValue::Alternatives(_, v) => merge_if_known(v, JsValue::is_string),
+
+            JsValue::Call(
+                _,
+                box JsValue::WellKnownFunction(
+                    WellKnownFunctionKind::RequireResolve
+                    | WellKnownFunctionKind::PathJoin
+                    | WellKnownFunctionKind::PathResolve(..)
+                    | WellKnownFunctionKind::OsArch
+                    | WellKnownFunctionKind::OsPlatform
+                    | WellKnownFunctionKind::PathDirname
+                    | WellKnownFunctionKind::PathToFileUrl
+                    | WellKnownFunctionKind::ProcessCwd,
+                ),
+                _,
+            ) => Some(true),
+
+            JsValue::FreeVar(..)
+            | JsValue::Variable(_)
+            | JsValue::Unknown(..)
+            | JsValue::Argument(..)
+            | JsValue::Call(..)
+            | JsValue::MemberCall(..)
+            | JsValue::Member(..) => None,
+        }
+    }
+
+    /// Checks if we known that the value starts with a given string. Returns
+    /// None if we don't know. Returns Some if we know if or if not the
+    /// value starts with the given string.
+    pub fn starts_with(&self, str: &str) -> Option<bool> {
```

Comment:
Might be good to give an example of why this particular function is needed, e.g. "we want to know whether a string represents an absolute path at compile-time".